load_farbfeld.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2021 Ricardo B. Ayres <ricardo.bosqueiro.ayres@gmail.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <pixman.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "functions.h"
+
+#define FF_MAGIC "farbfeld"
+
+
+/* using rob pike's byte order fallacy blog post */
+static uint32_t
+ffstol(uint8_t *data)
+{
+	return (data[3]<<0) | (data[2]<<8) | (data[1]<<16) | (data[0]<<24);
```

Comment:
Even though this code looks innocent, it introduces undefined C behaviour. shifting uin8_t by 24 bits could lead to a signed integer overflow if data[0] is 0x80. Signed overflow? Yes indeed! ;) The data[0] data type uin8_t is elevated to int by default, which is signed 32 bit.

You can see this if you prepare such a file and compile xwallpaper with gcc --fsanitize=undefined

Reply:
It is simpler to use ntohs as seen here: https://git.suckless.org/farbfeld/file/util.c.html#l67

Reply:
I suppose the same problem applies here, where I try to pack the bytes to write the pixel...
If I were to explicitly cast these bytes to appropriately sized unsigned types, would it solve the problem? (I don't plan on doing that, I'm just curious, really)

Reply:
Oh yes, I totally missed that spot! Thanks for highlighting it. And yes, casting to unsigned int or anything larger than int would solve the problem. You can see an example of this here: https://gitlab.freedesktop.org/xorg/lib/libxcursor/-/commit/204b6f130858ef038832887ea10488e7aed711a6